### DISCUSSION TITLE
How to verify SLEEF port on riscv64 platform

### DISCUSSION THREAD

**Discussion by soccercheng** [pocl/pocl#1550] [Q&A]
Does current pocl have any testing cases/programs for verifying SLEEF port for  riscv64 platform ?

**Comment by franz** [comment]
By "SLEEF port" you mean PoCL compiled with `-DENABLE_SLEEF=1` ? Since SLEEF is only used for math builtin functions, the usual way is to use OpenCL-CTS. However IIRC it does not compile on RISC-V currently, there were some architecture-dependent parts and nobody implemented those for RISC-V (yet).

**Reply by soccercheng** [reply]
Yes, I notice that.

The [current SLEEF](https://github.com/shibatch/sleef) code has better RVV1/RVV2 support,  so I try to integrate the current SLEEF implementation with POCL, I can successfully compile the code and build libpocl.so with SLEEF + RVV implementation. So, I need testing cases to verify my integration.

BTW, do you know why the current pocol SLEEF integration separates 64/128/256 vector bit length independently ? It seems that essential SLEEF does NOT do this way ?

**Reply by franz** [reply]
> So, I need testing cases to verify my integration.

I don't know of any math tests outside of OpenCL-CTS. Unfortunately calculating the reference values for verification is not that simple, you can take a look in the CTS code.

> SLEEF integration separates 64/128/256 vector bit length independently

Probably because PoCL (especially with distro build) has to support multiple generations of CPUs with different available vector register widths (128bit SSE / 256/512bit AVX), but it still has to implement builtin math functions for all OpenCL vector widths (up to 1024bit double16 / long16).

**Comment by soccercheng** [comment]
I am able to x-build conformance with essential POCL build flow and test an my Banana Pi BPI-F3 with riscv64 now, here are my basic system setup:

- Just a few patches to conformance for "BYPASS" on architecture-dependent parts, most of them are floating point related, like "Set the reference hardware floating point unit to FTZ mode", and I'll look into these issues and see if I can help on these.
- The testing target pocl runtime library is NOT built with "ENABLE_CONFORMANCE=ON", cause I want to see how bad it is when pocl is not built  with "ENABLE_CONFORMANCE=ON"

The followings are the FAILED tests on math builtin functions:

        226 - conformance_main_math_micro_acosh (Failed)
        238 - conformance_main_math_micro_copysign (Failed)
        254 - conformance_main_math_micro_frexp (Failed)
        271 - conformance_main_math_micro_lgamma (SEGFAULT)
        272 - conformance_main_math_micro_lgamma_r (Subprocess aborted)
        276 - conformance_main_math_micro_log1p (Failed)
        285 - conformance_main_math_micro_pown (Failed)
        288 - conformance_main_math_micro_remquo (SEGFAULT)
        290 - conformance_main_math_micro_rootn (Failed)

**Reply by pjaaskel** [reply]
Great progress! Looking forward to your pull requests to help PoCL improve on RISC-V.
